pkg/generator/gen.go

```diff
@@ -0,0 +1,340 @@
+/*
+Copyright The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generator
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"go/format"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/kcp-dev/client-gen/pkg/flag"
+	"github.com/kcp-dev/client-gen/pkg/internal"
+	"k8s.io/code-generator/cmd/client-gen/args"
+	"k8s.io/code-generator/cmd/client-gen/types"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/loader"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+)
+
+var (
+	// RuleDefinition is a marker for defining rules
+	RuleDefinition = markers.Must(markers.MakeDefinition("genclient", markers.DescribesType, placeholder{}))
+)
+
+const (
+	// output of the generated wrappers is written under <outputDir>/generated path.
+	generatedFolderName = "generated"
```

Comment:
Can we remove this constant and let the user supply it? Can it be part of outputDir?

Reply:
This is being added to the `outputDir` provided by user. The code-generator usually scaffolds the clients inside a predefined `clientset_generated/` dir and individual clients under `generated/` ([ref](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/generating-clientset.md#output-of-client-gen)). I am fine with either way if we still want it to be just outputDir provided by user.

Reply:
kcp's generated clientset is `pkg/client/clientset/versioned`, with subpackages under `versioned` such as `typed/tenancy/v1alpha1`.

client-go's is just `kubernetes`, with subpackages such as `typed/apps/v1`.

These come from passing (for kcp) `--clientset-name=versioned` and `--output-package github.com/kcp-dev/kcp/pkg/client/clientset`.

If you're ok with it, we could do 2 similar flags: 1 for clientset name, and 1 for output package. WDYT?